crossl2inbox/crossl2inbox.go

```diff
@@ -0,0 +1,134 @@
+package crossl2inbox
+
+import (
+	"bytes"
+	_ "embed"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-service/sources/batching"
+	"github.com/ethereum-optimism/supersim/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	ethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/holiman/uint256"
+)
+
+const (
+	eventExecutingMessage = "ExecutingMessage"
+)
+
+var (
+	ErrEventNotFound    = errors.New("event not found")
+	crossL2InboxAddress = common.HexToAddress("0x4200000000000000000000000000000000000022")
+)
+
+//go:embed CrossL2Inbox.json
+var crossL2Inbox []byte
+
+type ExecutingMessage struct {
+	Message    []byte
+	Identifier Identifier
+}
+
+// Identifier struct to match the Solidity struct
+type Identifier struct {
+	Origin      common.Address
+	BlockNumber *big.Int
+	LogIndex    *big.Int
+	Timestamp   *big.Int
+	ChainId     *big.Int
+}
+
+type CrossL2Inbox struct {
+	contract        *batching.BoundContract
+	crossL2InboxAbi abi.ABI
+}
+
+func NewCrossL2Inbox() *CrossL2Inbox {
+	abi := loadCrossL2InboxABI()
+	return &CrossL2Inbox{
+		crossL2InboxAbi: *abi,
+		contract:        batching.NewBoundContract(abi, crossL2InboxAddress),
+	}
+}
+
+func (i *CrossL2Inbox) DecodeExecutingMessageLog(l *ethTypes.Log) (*ExecutingMessage, error) {
+	if l.Address != i.contract.Addr() {
+		return nil, nil
+	}
+	name, result, err := i.contract.DecodeEvent(l)
+	if errors.Is(err, batching.ErrUnknownEvent) {
+		return nil, fmt.Errorf("%w: %v", ErrEventNotFound, err.Error())
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to decode event: %w", err)
+	}
+	if name != eventExecutingMessage {
+		return nil, nil
+	}
+
+	encodedId := result.GetBytes(0)
+	message := result.GetBytes(1)
+	identifierArgs := abi.Arguments{
```

Comment:
This library seems helpful here: https://github.com/lmittmann/w3/blob/v0.16.8/event.go#L26

https://github.com/lmittmann/w3?tab=readme-ov-file#abi-bindings

Reply:
It doesnt look like that library offers the ability to decode the encoded id inside the event into the Identifier struct?